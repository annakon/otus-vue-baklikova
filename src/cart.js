import { ref } from 'vue';

export const cart = ref([]);

export function addToCart(product) {
  let item = cart.value.find((i) => i.id === product.id);

  if (item) {
    item.quantity++;
  } else {
    cart.value.push({ ...product, quantity: 1 });
  }
}
