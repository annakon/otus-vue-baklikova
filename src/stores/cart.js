import { ref } from 'vue';
import { defineStore } from 'pinia';

export const useCartStore = defineStore('cart', () => {
  const cart = ref(new Map());
  const totalSum = ref(0);

  function addToCart(product) {
    let item = cart.value.get(product.id);

    if (item) {
      item.quantity++;
      totalSum.value = totalSum.value + product.price * item.quantity;
      alert('В корзине: ' + item.quantity);
    } else {
      cart.value.set(product.id, { ...product, quantity: 1 });
      totalSum.value = totalSum.value + product.price;
      alert('В корзине: 1');
    }
  }

  function clearCart() {
    cart.value.clear();
    totalSum.value = 0;
  }

  return { cart, addToCart, clearCart, totalSum };
});
