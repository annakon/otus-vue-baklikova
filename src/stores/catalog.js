import {reactive, ref} from 'vue';
import { defineStore } from 'pinia';
import axios from "axios";

export const useCatalogStore = defineStore('catalog', () => {
    const goods = reactive([]);
    const errored = ref();
    const loading = ref(true);
    const request = 'https://fakestoreapi.com/products';
    const product = ref();

    async function requestGoods() {
        await axios
            .get(request)
            .then((response) => (goods.push(...response.data)))
            .catch((error) => {
                console.log(error);
                errored.value = true;
            })
            .finally(() => (loading.value = false));
    }

    async function requestProduct(requestOneProduct) {
        await axios
            .get(requestOneProduct)
            .then((response) => (product.value = response.data))
            .catch((error) => {
                console.log(error);
                errored.value = true;
            })
            .finally(() => (loading.value = false));
    }

    return { goods,
        errored,
        loading,
        product,
        requestProduct,
        requestGoods,};
});
