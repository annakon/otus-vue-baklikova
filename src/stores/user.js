import { defineStore } from 'pinia';
import { computed, ref } from 'vue';
import axios from "axios";

export const useUserStore = defineStore('user', () => {
  const login = ref('');
  const name = ref('');
  const email = ref('');
  const address = ref('');
  async function saveOrder(orderData, callback) {
    await axios.post('https://reqbin.com/echo/post/json', JSON.stringify(orderData),{
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      }
    })
      .then((response) => console.log(JSON.stringify(response)))
      .then(callback)
      .catch((error) => {
        alert(error);
      });
  }
  function setUser() {
    login.value = localStorage.getItem('productLogin');
  }
  function logout() {
    localStorage.setItem('productLogin', '');
    login.value = '';
  }
  const isNotAuthorized = computed(() => login.value === '');
  const isAdmin = computed(() => login.value === 'admin');
  return {
    logout,
    saveOrder,
    setUser,
    isNotAuthorized,
    login,
    name,
    email,
    address,
    isAdmin
  };
});
