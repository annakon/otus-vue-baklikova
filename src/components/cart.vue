<template>
  <GoodsItemCart v-for="item in cart" :key="item.id" :cart="item"> </GoodsItemCart>
</template>

<script setup>
import { cart } from '@/cart';
import GoodsItemCart from '@/components/GoodsItemCart.vue';
</script>

<style scoped></style>
