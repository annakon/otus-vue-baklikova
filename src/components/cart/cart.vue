<template>
  <GoodsItemCart v-for="item in storeCart.cart" :key="item.id" :cart="item"> </GoodsItemCart>
  <button @click="storeCart.clearCart" class="btn btn-primary" type="button">
    Очистить корзину
  </button>
  <router-link to="/order" class="btn btn-primary">Оформить заказ</router-link>
</template>

<script setup>
import { useCartStore } from '@/stores/cart';
import GoodsItemCart from '@/components/cart/GoodsItemCart.vue';

const storeCart = useCartStore();
</script>

<style scoped>
button {
  margin-left: 2rem;
}
</style>
