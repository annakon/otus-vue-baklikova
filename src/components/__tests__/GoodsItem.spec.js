import {describe, it, expect, vi, beforeEach} from 'vitest';

import { mount} from '@vue/test-utils';
import component from '@/components/displayProducts/GoodsItem.vue';
import {createPinia, setActivePinia} from "pinia";
import {useCartStore} from "@/stores/cart";

vi.mock('vue-router')
describe('GoodsItem', () => {
  const item = {"id":1,"title":"Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops","price":109.95,"description":"Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday","category":"men's clothing","image":"https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg","rating":{"rate":3.9,"count":120}};
  let wrapper

  beforeEach(() => {
    setActivePinia(createPinia())
    wrapper = mount(component, { global: {
        stubs: ['RouterLink'],
      },props: { goods:item } });
  })
  it('renders properly', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });
  it("contain в корзину", () => {
    expect(wrapper.html()).toContain("В корзину");
  });
  it("renders all", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("call alert", async () => {
    const Control = wrapper.find('button');
    expect(Control.exists()).toBe(true);
    window.alert = vi.fn();
    await Control.trigger("click");
    expect(window.alert).toBeCalled();
  });
  it("add to cart ", async () => {
    const Control = wrapper.find('button');
    expect(Control.exists()).toBe(true);
    window.alert = vi.fn();
    await Control.trigger("click");
    const cart = useCartStore();
    expect(cart.cart.get(1)).toStrictEqual({...item,...{quantity:1}});
  });
});
