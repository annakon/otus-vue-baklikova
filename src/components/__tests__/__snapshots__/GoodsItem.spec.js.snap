// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GoodsItem > renders all 1`] = `
"<div data-v-f80bcfde=\\"\\" class=\\"card text-center\\"><img data-v-f80bcfde=\\"\\" class=\\"card-img-top\\" src=\\"https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg\\" alt=\\"Card image\\">
  <div data-v-f80bcfde=\\"\\" class=\\"cardBody\\">
    <h5 data-v-f80bcfde=\\"\\" class=\\"card-title\\">Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops</h5>
    <h6 data-v-f80bcfde=\\"\\" class=\\"card-subtitle mb-2 text-muted\\">109.95</h6>
    <p data-v-f80bcfde=\\"\\" class=\\"card-text\\">Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday</p>
    <p data-v-f80bcfde=\\"\\" class=\\"card-text text-muted\\"> rate: 3.9 count: 120</p>
    <section data-v-f80bcfde=\\"\\">
      <router-link-stub data-v-f80bcfde=\\"\\" to=\\"[object Object]\\" class=\\"btn btn-primary\\"></router-link-stub><button data-v-f80bcfde=\\"\\" class=\\"btn btn-primary\\" type=\\"button\\"> В корзину </button>
    </section>
  </div>
</div>"
`;
